sdk/src/Core/Amazon.Runtime/Internal/Auth/AWS4aSigner.cs

```diff
@@ -0,0 +1,594 @@
+
+                counterValue = 1;
+                while (counterValue < 0xFF)
+                {
+                    byte[] kDerived = ComputeKeyedHash(SignerAlgorithm, ksecret, kvalue);
+                    if (CompareConstantTime(kDerived, NMinus2) > 0)
+                    {
+                        // increment the counter value
+                        counterValue++;
+                        continue;
+                    }
+
+                    AddOneConstantTime(kDerived);
+                    var ecdsa = ECDsa.Create(new ECParameters
+                    {
+                        Curve = SigningCurve,
+                        D = kDerived,
+                    });
+                    Array.Clear(kDerived, 0, kDerived.Length);
+                    return ecdsa;
+                }
+
+                throw new AmazonClientException("Failed to derive a SigV4a key for the request.");
+            }
+            finally
+            {
+                // clean up all secrets, regardless of how initially seeded (for simplicity)
+                if (ksecret != null)
+                    Array.Clear(ksecret, 0, ksecret.Length);
+                if (kvalue != null)
+                    Array.Clear(kvalue, 0, kvalue.Length);
+            }
+        }
+
+        /// <summary>
+        /// Returns the ECDSA signature for an arbitrary blob using the specified credentials.
+        /// </summary>
+        /// <param name="credentials">The credentials to use.</param>
+        /// <param name="data">The data to sign.</param>
+        public static byte[] SignBlob(ImmutableCredentials credentials, string data)
+        {
+            return SignBlob(credentials, Encoding.UTF8.GetBytes(data));
+        }
+
+        /// <summary>
+        /// Returns the ECDSA signature for an arbitrary blob using the specified credentials.
+        /// </summary>
+        /// <param name="credentials">The credentials to use.</param>
+        /// <param name="data">The data to sign.</param>
+        public static byte[] SignBlob(ImmutableCredentials credentials, byte[] data)
+        {
+            var key = credentials.AWS4aSigningKey ??= ComputeSigningKey(credentials.AccessKey, credentials.SecretKey);
+#if NET7_0_OR_GREATER
+            return key.SignData(data, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence);
+#else
+            return ConvertToRfc3279DerSequence(key.SignData(data, HashAlgorithmName.SHA256));
+#endif
+        }
+
+#if !NET7_0_OR_GREATER
+        private static byte[] ConvertToRfc3279DerSequence(byte[] signature)
+        {
+            var writer = new AsnWriter(AsnEncodingRules.DER);
+            writer.PushSequence();
+            writer.WriteIntegerUnsigned(signature.AsSpan(0, signature.Length / 2)); // R value
+            writer.WriteIntegerUnsigned(signature.AsSpan(signature.Length / 2)); // S value
+            writer.PopSequence();
+            return writer.Encode();
+        }
+#endif
```

Comment:
The fact that we format the signature in ASN.1 is [not documented anywhere](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_sigv-create-signed-request.html), and stumped me for some days. I sent feedback to update the documentation.